keyspace/resourcegroup/resourcemanager.go

```diff
@@ -0,0 +1,121 @@
+func MakeRequestInfo(req *tikvrpc.Request) *RequestInfo {
+	if !req.IsTxnWriteRequest() && !req.IsRawWriteRequest() {
+		return &RequestInfo{writeBytes: -1}
+	}
+
+	var writeBytes int64
+	switch r := req.Req.(type) {
+	case *kvrpcpb.PrewriteRequest:
+		writeBytes += int64(r.TxnSize)
+	case *kvrpcpb.CommitRequest:
+		writeBytes += int64(unsafe.Sizeof(r.Keys))
+	}
+
+	return &RequestInfo{writeBytes: writeBytes}
+}
+
+// IsWrite returns whether the request is a write request.
+func (req *RequestInfo) IsWrite() bool {
+	return req.writeBytes > -1
+}
+
+// WriteBytes returns the actual write size of the request,
+// -1 will be returned if it's not a write request.
+func (req *RequestInfo) WriteBytes() uint64 {
+	return uint64(req.writeBytes)
+}
+
+// ResponseInfo contains information about a response that is able to calculate the RU cost
+// after the response is received. Specifically, the read bytes RU cost of a read request
+// could be calculated by its response size, and the KV CPU time RU cost of a request could
+// be calculated by its execution details info.
+type ResponseInfo struct {
+	readBytes uint64
+	kvCPUMs   uint64
+}
+
+// MakeResponseInfo extracts the relevant information from a BatchResponse.
+func MakeResponseInfo(resp *tikvrpc.Response) *ResponseInfo {
+	if resp.Resp == nil {
+		return &ResponseInfo{}
+	}
+	// Parse the response to extract the info.
+	var (
+		readBytes uint64
+		detailsV2 *kvrpcpb.ExecDetailsV2
+		details   *kvrpcpb.ExecDetails
+	)
+	switch r := resp.Resp.(type) {
+	case *coprocessor.Response:
+		detailsV2 = r.GetExecDetailsV2()
+		details = r.GetExecDetails()
+		readBytes = uint64(r.Data.Size())
+	case *tikvrpc.CopStreamResponse:
+		// Streaming request returns `io.EOF``, so the first `CopStreamResponse.Response`` may be nil.
+		if r.Response != nil {
+			detailsV2 = r.Response.GetExecDetailsV2()
+			details = r.Response.GetExecDetails()
+		}
+		readBytes = uint64(r.Data.Size())
+	case *kvrpcpb.GetResponse:
+		detailsV2 = r.GetExecDetailsV2()
+	case *kvrpcpb.BatchGetResponse:
+		detailsV2 = r.GetExecDetailsV2()
+	case *kvrpcpb.ScanResponse:
+		// TODO: using a more accurate size rather than using the whole response size as the read bytes.
+		readBytes = uint64(r.Size())
+	default:
+		// TODO: support the raw kv response maybe.
+		log.Warn("[kv resource] unknown response type to collect the info", zap.Any("type", reflect.TypeOf(r)))
+		return &ResponseInfo{}
+	}
+	// Try to get read bytes from the `detailsV2`.
+	// TODO: clarify whether we should count the underlying storage engine read bytes or not.
+	if scanDetail := detailsV2.GetScanDetailV2(); scanDetail != nil {
+		readBytes = scanDetail.GetProcessedVersionsSize()
+	}
+	// Get the KV CPU time in milliseconds from the execution time details.
+	kvCPUMs := getKVCPUMs(detailsV2, details)
+	return &ResponseInfo{readBytes: readBytes, kvCPUMs: kvCPUMs}
+}
+
+func getKVCPUMs(detailsV2 *kvrpcpb.ExecDetailsV2, details *kvrpcpb.ExecDetails) uint64 {
```

Comment:
I'm afraid wall time cpu is not so accurate here because it contains some wait time, so when the load is high, the same request will take longer to finish. @sticnarf Do you have any idea about this

Reply:
Truely. If we want to get the CPU time used by a request, `process_wall_time_ms` is inaccurate. It is reasonable to me to add another field for the real CPU cost.

Reply:
Will GetProcessWallTimeMs() - GetWaitWallTimeMs() good enough at present?

Reply:
No. For example, coprocessor may suspend a long request and execute other queries. The suspending time is counted in `PorcessWallTimeMs` but not counted in `WaitWallTimeMs`.

Reply:
Currently, he CPU precision does not need to be very high, Let's provide a better method later.